File: questPresentation/src/main/java/ru/nekit/android/qls/quest/types/TimeQuest.java
package ru.nekit.android.qls.quest.types;

import android.support.annotation.NonNull;

import ru.nekit.android.qls.quest.Quest;
import ru.nekit.android.qls.quest.QuestType;
import ru.nekit.android.qls.quest.QuestionType;
import ru.nekit.android.qls.quest.resources.common.IGroupWeightComparisonQuest;
import ru.nekit.android.qls.utils.MathUtils;

public class TimeQuest extends NumberSummandQuest implements IGroupWeightComparisonQuest {

    private int groupComparisonType;

    public TimeQuest(@NonNull Quest quest) {
        setQuestType(QuestType.TIME);
        setQuestionType(quest.getQuestionType());
        NumberSummandQuest inQuest = (NumberSummandQuest) quest;
        leftNode = inQuest.leftNode;
        int length = leftNode.length;
        if (quest.getQuestionType() == QuestionType.UNKNOWN_MEMBER) {
            unknownMemberIndex = MathUtils.randUnsignedInt(length - 1);
        }
        if (quest.getQuestionType() == QuestionType.COMPARISON) {
            groupComparisonType = MathUtils.randInt(MIN_GROUP_WEIGHT, MAX_GROUP_WEIGHT);
        }
    }

    public static int getTimeHours(int time) {
        return (time - time % 60) / 60;
    }

    public static int getTimeMinutes(int time) {
        return time - getTimeHours(time) * 60;
    }

    private int getHoursByIndex(int index) {
        int time = leftNode[index];
        return (time - time % 60) / 60;
    }

    private int getMinutesByIndex(int index) {
        int time = leftNode[index];
        return time - getHoursByIndex(index) * 60;
    }

    public int getUnknownTime() {
        return getUnknownMember();
    }

    public int getUnknownTimeHours() {
        return getHoursByIndex(unknownMemberIndex);
    }

    public int getUnknownTimeMinutes() {
        return getMinutesByIndex(unknownMemberIndex);
    }

    private String getTimeString(int index) {
        int hours = getHoursByIndex(index);
        int minutes = getMinutesByIndex(index);
        return String.format("%s:%s", hours, minutes < 10 ? "0" + minutes : minutes);
    }

    public String getUnknownTimeString() {
        return getTimeString(unknownMemberIndex);
    }

    @Override
    public int getGroupComparisonType() {
        return getQuestionType() == QuestionType.COMPARISON ? groupComparisonType : -1;
    }

    @Override
    public Integer getAnswer() {
        if (getQuestionType() == QuestionType.COMPARISON) {
            boolean isMax = getGroupComparisonType() == MAX_GROUP_WEIGHT;
            int answer = isMax ? 0 : Integer.MAX_VALUE;
            for (int item : leftNode) {
                answer = isMax ? Math.max(answer, item) : Math.min(answer, item);
            }
            return answer;
        }
        return super.getAnswer();
    }
}

File: questDomain/src/main/java/ru/nekit/android/qls/domain/model/quest/TimeQuest.kt
package ru.nekit.android.qls.domain.model.quest

import ru.nekit.android.qls.domain.model.resources.common.IGroupWeightComparisonQuest
import ru.nekit.android.qls.domain.model.resources.common.IGroupWeightComparisonQuest.Companion.MAX_GROUP_WEIGHT
import ru.nekit.android.qls.shared.model.QuestionType.COMPARISON
import ru.nekit.android.qls.shared.model.QuestionType.UNKNOWN_MEMBER
import ru.nekit.android.utils.MathUtils.randInt
import ru.nekit.android.utils.MathUtils.randUnsignedInt

open class TimeQuest(quest: NumberSummandQuest) : NumberSummandQuest(), IGroupWeightComparisonQuest {

    final override var groupComparisonType: Int = -1

    val unknownTime: Int
        get() = unknownMember

    val unknownTimeHours: Int
        get() = getHoursByIndex(unknownMemberIndex)

    val unknownTimeMinutes: Int
        get() = getMinutesByIndex(unknownMemberIndex)

    override val answer: Int
        get() {
            if (questionType == COMPARISON) {
                return (if (groupComparisonType == MAX_GROUP_WEIGHT)
                    leftNode.max()
                else
                    leftNode.min())!!
            }
            return super.answer
        }

    init {
        leftNode = quest.leftNode
        val length = leftNode.size
        if (quest.questionType == UNKNOWN_MEMBER) {
            unknownMemberIndex = randUnsignedInt(length - 1)
        } else if (quest.questionType == COMPARISON) {
            groupComparisonType = randInt(IGroupWeightComparisonQuest.MIN_GROUP_WEIGHT,
                    MAX_GROUP_WEIGHT)
        }
    }

    fun getHoursByIndex(index: Int): Int {
        val time = leftNode[index]
        return (time - time % TIME_METRICS) / TIME_METRICS
    }

    fun getMinutesByIndex(index: Int): Int {
        val time = leftNode[index]
        return time - getHoursByIndex(index) * TIME_METRICS
    }

    companion object {
        fun getTimeHours(time: Int): Int {
            return (time - time % TIME_METRICS) / TIME_METRICS
        }

        fun getTimeMinutes(time: Int): Int {
            return time - getTimeHours(time) * TIME_METRICS
        }

        val TIME_METRICS = 60

    }
}